sklearn/utils/tests/test_estimator_checks.py

```diff
@@ -571,6 +582,11 @@ def test_check_estimator():
     # doesn't error on binary_only tagged estimator
     check_estimator(TaggedBinaryClassifier())
 
+    # Check regressor with requires_positive_X estimator tag
+    msg = "negative X values not supported!"
+    with raises(ValueError, match=msg):
+        check_estimator(RequiresPositiveXRegressor())
```

Comment:
I think the tests for `RequiresPositiveYRegressor` and `RequiresPositiveXRegressor`, are asserting the incorrect behavior. If an estimator is tagged with `"requires_positive_X": True`, I expect `check_estimator` would generate the correct data and not fail `check_estimator`. 

For this PR, let's focus on `RequiresPositiveXRegressor` by making sure the following passes:

```python
check_estimator(RequiresPositiveXRegressor())
```

This means inserting `_enforce_estimator_tags_x` in a few more places in `estimator_check`.

Also, `_enforce_estimator_tags_x` currently includes zero when generating `X`:

https://github.com/scikit-learn/scikit-learn/blob/f09f8dd982d59a6e51eca65616ad9f97d4000561/sklearn/utils/estimator_checks.py#L3273-L3274

This is different from `y` which does not include the zero. For me, I think it's better to update the behavior to how `y` does it and not include 0 when generating data. This is a more lenient definition for third party estimators.